metaflow/plugins/aws/awslambda/lambda_decorator.py

```diff
@@ -0,0 +1,213 @@
+from metaflow.plugins.aws.awslambda.lambda_runner import LAMBDA_MAX_MEMORY_MB
+import sys
+from typing import TYPE_CHECKING, Optional, List, Dict
+from metaflow.decorators import StepDecorator, Decorator
+from metaflow.plugins.timeout_decorator import get_run_time_limit_for_task
+from metaflow import R
+from metaflow.datastore.datastore import TransformableObject
+from metaflow.exception import MetaflowException
+from metaflow.plugins.aws.sync_metadata import sync_metadata_to_s3
+from metaflow.plugins.aws.batch.batch_decorator import (
+    BatchDecorator,
+    DEFAULT_TIMEOUT_SECONDS,
+    ResourcesDecorator,
+)
+from metaflow.plugins.conda.conda_step_decorator import CondaStepDecorator
+
+if TYPE_CHECKING:
+    from metaflow.runtime import CLIArgs
+    from metaflow.graph import FlowGraph
+    from metaflow.package import MetaflowPackage
+    from metaflow.flowspec import FlowSpec
+    from metaflow.datastore import MetaflowDataStore
+    from metaflow.environment import MetaflowEnvironment
+    from metaflow.event_logger import EventLogger
+
+from metaflow.metaflow_config import (
+    LAMBDA_LOG_GROUP_NAME,
+    LAMBDA_ROLE_ARN,
+    LAMBDA_IMAGE_URI,
+)
+
+# Supported resource specs for Lambda
+LAMBDA_MAX_RESOURCES = {
+    "memory": LAMBDA_MAX_MEMORY_MB,
+    "cpu": 6,
+}
+DEFAULT_TIMEOUT_SECONDS = 600
+
+
+class LambdaException(MetaflowException):
+    headline = "AWS Lambda error"
+
+
+class LambdaDecorator(StepDecorator):
+    """
+    The decorator is called @awslambda to avoid confusion, since "lambda" is a reserved
+    word in Python.
+    """
+
+    name = "awslambda"
+    defaults = {
+        "image": None,
+        "force_fetch_image": False,
+    }
+
+    package_url: Optional[str] = None
+    package_sha: Optional[str] = None
+    resource_attributes: Dict[str, str]
+    run_time_limit = None
+    ds_root: Optional[str] = None
+
+    def step_init(
+        self,
+        flow: "FlowSpec",
+        graph: "FlowGraph",
+        step: str,
+        decos: List[Decorator],
+        environment: "MetaflowEnvironment",
+        datastore: "MetaflowDataStore",
+        logger: "EventLogger",
+    ):
+        if datastore.TYPE != "s3":
+            raise LambdaException("The *@awslambda* decorator requires --datastore=s3.")
+
+        self.logger = logger
+        self.environment = environment
+        self.step = step
+        self.resource_attributes = {}
+        for deco in decos:
+            if isinstance(deco, CondaStepDecorator):
+                raise LambdaException(
```

Comment:
What would it take to support `@conda` in `@awslambda`?

Reply:
One idea (mentioned it briefly in the PR decription) is to build custom docker images from conda envs, push them to ECR and use them to create Lambdas.

Reply:
Maybe, we can consider using [lambda layers](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html) rather than building images on the fly - which will presumably need more infrastructure. That can be a follow-on PR.